File: packages/cli/hieudoanm/cli.kt/src/main/kotlin/io/github/hieudoanm/cli/commands/Web.kt
package io.github.hieudoanm.cli.commands

import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.subcommands
import com.github.ajalt.clikt.parameters.arguments.argument
import com.github.ajalt.clikt.parameters.arguments.default
import com.github.ajalt.clikt.parameters.options.option
import com.github.ajalt.clikt.parameters.options.flag
import com.github.ajalt.clikt.parameters.options.default
import com.github.ajalt.clikt.parameters.options.required
import com.github.ajalt.clikt.parameters.types.int
import com.google.gson.GsonBuilder
import io.github.hieudoanm.cli.services.Requests
import java.io.File
import java.net.URL

class WebCommand : CliktCommand(name = "web", help = "Web service tools") {
    init {
        subcommands(
            WebInstagram(), WebShopify(), WebSnapshot(), WebWeather(), WebYoutube()
        )
    }
    override fun run() = Unit
}

class WebInstagram : CliktCommand(name = "instagram", help = "Instagram related tools") {
    init { subcommands(WebInstagramDownload()) }
    override fun run() = Unit
}

class WebInstagramDownload : CliktCommand(name = "download", help = "Download images from Instagram") {
    private val url by option("--url", "-u", help = "Instagram post URL or shortcode").required()
    private val output by option("--output", "-o", help = "Output directory").default(".")
    private val index by option("--index", "-i", help = "Specific image index (1-based)").int().default(0)
    private val proxy by option("--proxy", "-p", help = "Use proxy").flag()
    private val json by option("--json", help = "Output in JSON format").flag()

    override fun run() {
        val shortcode = extractShortcode(url)
        val html = fetchIGPage(shortcode, proxy)
        val imageURLs = scrapeIGImages(html)
        if (imageURLs.isEmpty()) {
            echo("no images found in post $shortcode")
            return
        }

        if (json) {
            val gson = GsonBuilder().setPrettyPrinting().create()
            echo(gson.toJson(mapOf("shortcode" to shortcode, "image_urls" to imageURLs, "count" to imageURLs.size)))
            return
        }

        echo("Shortcode: $shortcode")
        echo("Found ${imageURLs.size} image candidates")

        val targets = if (index > 0) {
            if (index > imageURLs.size) throw IllegalArgumentException("index $index out of range (found ${imageURLs.size} images)")
            echo("Selected image index $index")
            listOf(imageURLs[index - 1])
        } else imageURLs

        File(output).mkdirs()
        var downloaded = 0
        targets.forEachIndexed { i, imgUrl ->
            val filename = if (targets.size == 1 && index > 0) "${shortcode}_$index.jpg"
                else if (targets.size == 1) "$shortcode.jpg"
                else "${shortcode}_${i + 1}.jpg"
            echo("  Downloading $filename ... ", trailingNewline = false)
            try {
                val url = URL(imgUrl)
                url.openStream().use { input -> File(File(output), filename).outputStream().use { it.write(input.readBytes()) } }
                echo("saved -> $filename")
                downloaded++
            } catch (e: Exception) {
                echo("failed: ${e.message}")
            }
        }
        echo("Done. $downloaded image(s) saved to $output")
    }
}

private val igShortcodePattern = Regex("(?:instagram\\.com/(?:p|reels|tv|reel)/)([^/?#&]+)")
private val igDisplayUrlPattern = Regex("\"display_url\":\"([^\"]+)\"")
private val igOgImagePattern = Regex("<meta property=\"og:image\" content=\"([^\"]+)\"")

private fun extractShortcode(input: String): String {
    val m = igShortcodePattern.find(input.trim())
    if (m != null) return m.groupValues[1]
    if (!input.contains("/") && input.trim().length >= 10) return input.trim()
    throw IllegalArgumentException("could not extract Instagram shortcode from: $input")
}

private fun fetchIGPage(shortcode: String, useProxy: Boolean): String {
    val targetURL = "https://www.instagram.com/p/$shortcode/"
    if (useProxy) {
        val proxyURL = "https://hieudoanm-proxy.vercel.app/api?url=${java.net.URLEncoder.encode(targetURL, "UTF-8")}"
        val result = Requests.get(proxyURL)
        return result.getOrThrow()
    }
    val result = Requests.get(targetURL, mapOf(
        "User-Agent" to "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36"
    ))
    return result.getOrThrow()
}

private fun scrapeIGImages(html: String): List<String> {
    val urls = mutableListOf<String>()
    val seen = mutableSetOf<String>()
    for (m in igDisplayUrlPattern.findAll(html)) {
        val u = m.groupValues[1].replace("\\u0026", "&")
        if (seen.add(u)) urls.add(u)
    }
    if (urls.isEmpty()) {
        val m = igOgImagePattern.find(html)
        if (m != null) urls.add(m.groupValues[1])
    }
    return urls
}

class WebShopify : CliktCommand(name = "shopify", help = "Shopify detection and analysis tools") {
    init { subcommands(WebShopifyDetect()) }
    override fun run() = Unit
}

class WebShopifyDetect : CliktCommand(name = "detect", help = "Detect if a website is using Shopify") {
    private val url by argument().default("")
    private val verbose by option("--verbose", "-v", help = "Show detection signals").flag()
    private val json by option("--json", help = "Output in JSON format").flag()

    override fun run() {
        var target = url
        if (target.isEmpty()) {
            echo("Enter website URL: ")
            target = readLine()?.trim() ?: ""
        }
        if (target.isEmpty()) throw IllegalArgumentException("URL is required")
        if (!target.startsWith("http://") && !target.startsWith("https://")) target = "https://$target"

        val (isShopify, isPlus, signals) = checkShopify(target)
        if (json) {
            val gson = GsonBuilder().setPrettyPrinting().create()
            echo(gson.toJson(mapOf("url" to target, "shopify" to isShopify, "plus" to isPlus, "signals" to signals)))
            return
        }
        if (isShopify) echo("Shopify detected")
        else {
            echo("Shopify not detected")
            return
        }
        if (isPlus) echo("Shopify Plus detected")
        else echo("Shopify Plus not detected")
        if (verbose && signals.isNotEmpty()) {
            echo("Signals:")
            signals.forEach { echo(" - $it") }
        }
    }
}

private data class ShopifyResult(val isShopify: Boolean, val isPlus: Boolean, val signals: List<String>)

private fun checkShopify(urlStr: String): ShopifyResult {
    var isShopify = false
    var isPlus = false
    val signals = mutableListOf<String>()

    try {
        val url = URL(urlStr)
        val conn = url.openConnection() as java.net.HttpURLConnection
        conn.setRequestProperty("User-Agent", "shopify-check/1.0")
        conn.connectTimeout = 10000
        conn.readTimeout = 10000
        conn.instanceFollowRedirects = true

        for ((key, values) in conn.headerFields) {
            if (key == null) continue
            val lk = key.lowercase()
            if (lk.startsWith("x-shopify")) {
                isShopify = true
                signals.add("Header: $key")
                if (lk == "x-shopify-stage" || lk == "x-shopify-shop-api-call-limit") isPlus = true
            }
            values.forEach { v ->
                if (v.lowercase().contains("shopify")) {
                    isShopify = true
                    signals.add("Header value: $key")
                }
            }
        }

        val body = conn.inputStream.bufferedReader().readText().take(16 * 1024).lowercase()
        if (body.contains("cdn.shopify.com")) { isShopify = true; signals.add("HTML: cdn.shopify.com") }
        if (body.contains("shopify-section")) { isShopify = true; signals.add("HTML: shopify-section") }
        if (body.contains("shopify-plus")) { isPlus = true; signals.add("HTML: shopify-plus") }
    } catch (_: Exception) {}
    return ShopifyResult(isShopify, isPlus, signals)
}

class WebSnapshot : CliktCommand(name = "snapshot", help = "Take a screenshot of a web page") {
    private val url by option("--url", "-u", help = "URL to capture").required()
    private val output by option("--output", "-o", help = "Output file or directory").default("")
    private val width by option("--width", help = "Viewport width").int().default(0)
    private val height by option("--height", help = "Viewport height").int().default(0)
    private val preset by option("--preset", help = "Viewport preset: desktop|laptop|tablet|mobile|hd|4k").default("desktop")
    private val fullPage by option("--full-page", help = "Capture full scrollable page").flag()
    private val delay by option("--delay", help = "Wait before capturing (e.g. 500ms, 2s)").default("")
    private val pdf by option("--pdf", help = "Save as PDF instead of PNG").flag()
    private val quality by option("--quality", help = "Screenshot quality 1-100").int().default(90)
    private val verbose by option("--verbose", "-v", help = "Print extra info").flag()
    private val json by option("--json", help = "Output in JSON format").flag()

    override fun run() {
        var rawURL = url
        if (!rawURL.startsWith("http://") && !rawURL.startsWith("https://")) rawURL = "https://$rawURL"
        val ext = if (pdf) ".pdf" else ".png"
        val outPath = if (output.isNotEmpty()) output else "${System.getProperty("user.home")}/.snapshot/${hostnameSlug(rawURL)}_${java.time.LocalDateTime.now().format(java.time.format.DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"))}$ext"

        echo("Snapshot: Opening $rawURL in default browser (external tool)")
        try {
            val os = System.getProperty("os.name").lowercase()
            val cmd = when {
                os.contains("mac") -> arrayOf("open", rawURL)
                os.contains("linux") -> arrayOf("xdg-open", rawURL)
                else -> arrayOf("cmd", "/c", "start", rawURL)
            }
            Runtime.getRuntime().exec(cmd)
        } catch (e: Exception) {
            echo("Failed to open browser: ${e.message}")
        }

        if (json) {
            val gson = GsonBuilder().setPrettyPrinting().create()
            echo(gson.toJson(mapOf("url" to rawURL, "format" to ext.uppercase().removePrefix("."))))
        }
    }
}

private fun hostnameSlug(rawURL: String): String {
    return try {
        val u = URL(rawURL)
        u.host.replace(".", "_").replace(":", "_")
    } catch (_: Exception) { "snapshot" }
}

class WebWeather : CliktCommand(name = "weather", help = "Check current weather for a city") {
    private val city by argument().default("")
    private val forecast by option("--forecast", "-f", help = "Show 3-day forecast").flag()
    private val json by option("--json", help = "Output in JSON format").flag()
    private val units by option("--units", "-u", help = "Units: metric, imperial, uk").default("metric")

    override fun run() {
        var location = city
        if (location.isEmpty() && !forecast) {
            val ipResult = Requests.get("https://ipinfo.io/json")
            ipResult.onSuccess { body ->
                try {
                    val data = GsonBuilder().create().fromJson(body, Map::class.java)
                    location = data["city"]?.toString() ?: ""
                } catch (_: Exception) {}
            }
        }
        if (location.isEmpty()) throw IllegalArgumentException("provide a city name")

        val unitParam = when (units) {
            "us", "imperial" -> "u"
            "uk" -> "M"
            else -> "m"
        }
        val encoded = java.net.URLEncoder.encode(location, "UTF-8")
        val url = if (forecast) {
            if (json) "https://wttr.in/$encoded?format=j1&lang=en&$unitParam"
            else "https://wttr.in/$encoded?0&lang=en&$unitParam"
        } else {
            if (json) "https://wttr.in/$encoded?format=j1&lang=en&$unitParam"
            else "https://wttr.in/$encoded?format=%C+%t+%w+%h&$unitParam"
        }

        val result = Requests.get(url)
        result.onSuccess { body ->
            if (json) {
                try {
                    val data = GsonBuilder().create().fromJson(body, Any::class.java)
                    val gson = GsonBuilder().setPrettyPrinting().create()
                    echo(gson.toJson(data))
                } catch (_: Exception) { echo(body) }
            } else {
                echo(body.trim())
            }
        }.onFailure { echo("fetch error: ${it.message}") }
    }
}

class WebYoutube : CliktCommand(name = "youtube", help = "YouTube transcript and thumbnail tools") {
    init { subcommands(WebYoutubeThumbnails(), WebYoutubeTranscript()) }
    override fun run() = Unit
}

private data class YtQuality(val id: String, val label: String, val resolution: String)

private val ytQualities = listOf(
    YtQuality("maxresdefault", "Max Resolution", "1280x720"),
    YtQuality("sddefault", "SD", "640x480"),
    YtQuality("hqdefault", "HQ", "480x360"),
    YtQuality("mqdefault", "MQ", "320x180"),
    YtQuality("default", "Default", "120x90"),
    YtQuality("0", "Frame 0", "480x360"),
    YtQuality("1", "Frame 1", "120x90"),
    YtQuality("2", "Frame 2", "120x90"),
    YtQuality("3", "Frame 3", "120x90"),
)

private val ytVideoIDPatterns = listOf(
    Regex("(?:youtube\\.com/watch\\?.*v=)([a-zA-Z0-9_-]{11})"),
    Regex("(?:youtu\\.be/)([a-zA-Z0-9_-]{11})"),
    Regex("(?:youtube\\.com/embed/)([a-zA-Z0-9_-]{11})"),
    Regex("(?:youtube\\.com/shorts/)([a-zA-Z0-9_-]{11})"),
    Regex("^([a-zA-Z0-9_-]{11})$"),
)

private fun ytExtractVideoID(input: String): String {
    val trimmed = input.trim()
    for (re in ytVideoIDPatterns) {
        val m = re.find(trimmed)
        if (m != null) return m.groupValues[1]
    }
    throw IllegalArgumentException("could not extract video ID from: $input")
}

private fun ytThumbURL(videoID: String, qualityID: String): String =
    "https://img.youtube.com/vi/$videoID/$qualityID.jpg"

private fun ytValidQualityIDs(): String = ytQualities.joinToString(", ") { it.id }

private fun ytDownloadThumb(videoID: String, qualityID: String, outDir: String): String {
    val rawURL = ytThumbURL(videoID, qualityID)
    val url = URL(rawURL)
    val bytes = url.openStream().readBytes()
    if (bytes.size < 1000) throw IllegalArgumentException("quality $qualityID not available (too small: ${bytes.size} bytes)")
    val filename = "$videoID-$qualityID.jpg"
    val path = File(outDir, filename)
    path.writeBytes(bytes)
    return path.absolutePath
}

class WebYoutubeThumbnails : CliktCommand(name = "thumbnails", help = "Download YouTube video thumbnails") {
    private val url by option("--url", "-u", help = "Video URL or ID").required()
    private val quality by option("--quality", "-q", help = "Specific quality to download").default("")
    private val output by option("--output", "-o", help = "Output directory").default(".")
    private val all by option("--all", "-a", help = "Download all quality variants").flag()
    private val list by option("--list", "-l", help = "List thumbnail URLs without downloading").flag()
    private val json by option("--json", help = "Output in JSON format (with --list)").flag()

    override fun run() {
        val videoID = ytExtractVideoID(url)

        if (list) {
            if (json) {
                val entries = ytQualities.map { q -> mapOf("id" to q.id, "resolution" to q.resolution, "url" to ytThumbURL(videoID, q.id)) }
                val gson = GsonBuilder().setPrettyPrinting().create()
                echo(gson.toJson(mapOf("video_id" to videoID, "thumbnails" to entries)))
            } else {
                echo("Video ID: $videoID")
                echo("")
                echo("Available thumbnail URLs:")
                ytQualities.forEach { q -> echo("  ${"%-18s".format(q.id)} ${"%-12s".format(q.resolution)} ${ytThumbURL(videoID, q.id)}") }
            }
            return
        }

        echo("Video ID: $videoID")
        File(output).mkdirs()
        val targets = when {
            all -> ytQualities
            quality.isNotEmpty() -> {
                val found = ytQualities.firstOrNull { it.id == quality }
                    ?: throw IllegalArgumentException("unknown quality $quality — valid values: ${ytValidQualityIDs()}")
                listOf(found)
            }
            else -> ytQualities.take(3)
        }
        var downloaded = 0
        for (q in targets) {
            echo("  Downloading ${"%-18s".format(q.id)} (${q.resolution}) ... ", trailingNewline = false)
            try {
                val path = ytDownloadThumb(videoID, q.id, output)
                echo("saved -> $path")
                downloaded++
            } catch (e: Exception) {
                echo("skipped (${e.message})")
            }
        }
        echo("Done. $downloaded thumbnail(s) saved to $output")
    }
}

class WebYoutubeTranscript : CliktCommand(name = "transcript", help = "Fetch YouTube video transcript") {
    private val url by option("--url", "-u", help = "Video URL or ID").required()
    private val lang by option("--lang", "-l", help = "Language code (e.g. en, es, fr)").default("en")
    private val output by option("--output", "-o", help = "Save to file instead of stdout").default("")
    private val format by option("--format", "-f", help = "Output format: text or json").default("text")
    private val noTimestamps by option("--no-timestamps", help = "Omit timestamps from text output").flag()

    override fun run() {
        val videoID = ytExtractVideoID(url)
        val apiURL = "https://youtubetranscript.com/?v=$videoID&lang=$lang"
        val result = Requests.get(apiURL)
        result.onSuccess { body ->
            val out = body
            if (output.isNotEmpty()) File(output).writeText(out)
            else echo(out)
        }.onFailure { echo("Failed to fetch transcript: ${it.message}") }
    }
}

File: packages/cli/hieudoanm/swift/Sources/hieudoanm/cmd/web/web.swift
import Foundation
import ArgumentParser

struct WebCommand: ParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "web",
        abstract: "Web services",
        subcommands: [
            WebInstagram.self,
            WebShopify.self,
            WebSnapshot.self,
            WebWeather.self,
            WebYoutube.self,
        ]
    )
}

struct WebInstagram: ParsableCommand {
    static let configuration = CommandConfiguration(commandName: "instagram", abstract: "Download Instagram media from shortcode")

    @Argument(help: "Instagram URL or shortcode")
    var input: String

    @Option(name: .shortAndLong, help: "Output directory")
    var output: String = "."

    mutating func run() async throws {
        let shortcode: String
        if input.contains("/") {
            guard let sc = extractInstagramShortcode(from: input) else {
                print("Invalid Instagram URL")
                return
            }
            shortcode = sc
        } else {
            shortcode = input
        }

        let apiURL = "https://www.instagram.com/p/\(shortcode)/?__a=1"
        var imgURL: String?

        let resp = try? await Requests.fetch(apiURL, headers: ["User-Agent": "Mozilla/5.0", "Accept": "application/json"])
        if let data = resp?.data {
            let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            let items = (json?["items"] as? [[String: Any]]) ?? (json?["graphql"] as? [String: Any]).map { [$0] } ?? []
            if let item = items.first {
                imgURL = item["display_url"] as? String
                    ?? ((item["image_versions2"] as? [String: Any])?["candidates"] as? [[String: Any]])?.first?["url"] as? String
            }
        }

        if imgURL == nil {
            let htmlResp = try? await Requests.fetch("https://www.instagram.com/p/\(shortcode)/", headers: ["User-Agent": "Mozilla/5.0"])
            if let htmlData = htmlResp?.data, let htmlStr = String(data: htmlData, encoding: .utf8) {
                if let ogRange = htmlStr.range(of: "og:image"),
                   let contentStart = htmlStr[ogRange.upperBound...].range(of: "content=\""),
                   let end = htmlStr[contentStart.upperBound...].range(of: "\"") {
                    imgURL = String(htmlStr[contentStart.upperBound..<end.lowerBound])
                }
            }
        }

        guard let url = imgURL, !url.isEmpty else {
            print("Could not extract image URL")
            return
        }
        let imgResp = try await Requests.fetch(url)
        let filename = "\(output)/instagram_\(shortcode).jpg"
        try imgResp.data.write(to: URL(fileURLWithPath: filename))
        print("Downloaded to \(filename)")
    }
}

private func extractInstagramShortcode(from url: String) -> String? {
    let patterns = [
        try! NSRegularExpression(pattern: "instagram\\.com/p/([^/?]+)", options: []),
        try! NSRegularExpression(pattern: "instagram\\.com/reel/([^/?]+)", options: []),
        try! NSRegularExpression(pattern: "instagr\\.am/p/([^/?]+)", options: []),
        try! NSRegularExpression(pattern: "instagr\\.am/reel/([^/?]+)", options: []),
    ]
    let nsRange = NSRange(url.startIndex..<url.endIndex, in: url)
    for pattern in patterns {
        if let match = pattern.firstMatch(in: url, options: [], range: nsRange),
           let range = Range(match.range(at: 1), in: url) {
            return String(url[range])
        }
    }
    return nil
}

struct WebShopify: ParsableCommand {
    static let configuration = CommandConfiguration(commandName: "shopify", abstract: "Detect if a URL is a Shopify store")

    @Argument(help: "URL to check")
    var url: String

    mutating func run() async throws {
        let result = try await detectShopify(url: url)
        if result.isShopify {
            print("\(url) is using Shopify\(result.isPlus ? " Plus" : "")")
            print("Detected via: \(result.via)")
        } else {
            print("\(url) does not appear to be using Shopify")
        }
    }
}

struct ShopifyResult {
    let isShopify: Bool
    let isPlus: Bool
    let storeName: String?
    let via: String
}

private func detectShopify(url: String) async throws -> ShopifyResult {
    let resp = try await Requests.fetch(url, headers: ["User-Agent": "Mozilla/5.0"])
    let body = String(data: resp.data, encoding: .utf8) ?? ""
    if body.contains("x-shopify-plus") || body.contains("cdn.shopify.com") {
        return ShopifyResult(isShopify: true, isPlus: body.contains("x-shopify-plus"), storeName: nil, via: "html")
    }
    if body.contains("shopify-section") || body.contains("shopify-digital-wallet") {
        return ShopifyResult(isShopify: true, isPlus: body.contains("shopify-plus"), storeName: nil, via: "html")
    }
    return ShopifyResult(isShopify: false, isPlus: false, storeName: nil, via: "none")
}

struct WebSnapshot: ParsableCommand {
    static let configuration = CommandConfiguration(commandName: "snapshot", abstract: "Web page screenshot (opens in Safari)")

    @Argument(help: "URL to capture")
    var url: String

    @Option(name: .shortAndLong, help: "Output file")
    var output: String = "snapshot.png"

    mutating func run() async throws {
        try await Browser.screenshot(url: url, output: output)
    }
}

struct WebWeather: ParsableCommand {
    static let configuration = CommandConfiguration(commandName: "weather", abstract: "Show weather for a location")

    @Argument(help: "Location (city name or coordinates)")
    var location: String

    @Option(name: .shortAndLong, help: "Format: 1-4 (see wttr.in)")
    var format: Int = 1

    mutating func run() async throws {
        let query = location.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? location
        let url = "https://wttr.in/\(query)?format=\(format)"
        let resp = try await Requests.fetch(url, headers: ["User-Agent": "curl/7.68.0"])
        print(String(data: resp.data, encoding: .utf8) ?? "No data")
    }
}

struct WebYoutube: ParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "youtube",
        abstract: "YouTube tools",
        subcommands: [WebYoutubeThumbnails.self, WebYoutubeTranscript.self]
    )
}

struct WebYoutubeThumbnails: ParsableCommand {
    static let configuration = CommandConfiguration(commandName: "thumbnails", abstract: "Download YouTube thumbnails")

    @Argument(help: "YouTube video URL or ID")
    var video: String

    @Option(name: .shortAndLong, help: "Output directory")
    var output: String = "."

    @Option(name: .long, help: "Quality (maxresdefault, hqdefault, mqdefault, sddefault, default)")
    var quality: String?

    @Flag(name: .long, help: "Download all qualities")
    var all = false

    mutating func run() async throws {
        let videoID = extractVideoID(video)
        let qualities = ["maxresdefault", "hqdefault", "mqdefault", "sddefault", "default"]
        let selected = quality.map { [$0] } ?? (all ? qualities : ["maxresdefault"])
        for q in selected {
            let url = "https://img.youtube.com/vi/\(videoID)/\(q).jpg"
            do {
                let resp = try await Requests.fetch(url)
                let filename = "\(output)/\(videoID)_\(q).jpg"
                try resp.data.write(to: URL(fileURLWithPath: filename))
                print("Downloaded \(filename)")
            } catch {
                print("  \(q): not available")
            }
        }
    }
}

struct WebYoutubeTranscript: ParsableCommand {
    static let configuration = CommandConfiguration(commandName: "transcript", abstract: "Fetch YouTube transcript")

    @Argument(help: "YouTube video URL or ID")
    var video: String

    @Option(name: .shortAndLong, help: "Language code")
    var language: String = "en"

    @Flag(name: .long, help: "JSON output with timestamps")
    var json = false

    mutating func run() async throws {
        let videoID = extractVideoID(video)
        let transcript = try await fetchTranscript(videoID: videoID, language: language)
        if json {
            let lines = transcript.lines.map { l -> [String: Any] in
                ["text": l.text, "offset": l.offset, "duration": l.duration]
            }
            let data = try JSONSerialization.data(withJSONObject: ["transcript": lines], options: .prettyPrinted)
            print(String(data: data, encoding: .utf8) ?? "")
        } else {
            for line in transcript.lines {
                let timestamp = formatTime(line.offset)
                print("[\(timestamp)] \(line.text)")
            }
        }
    }
}

private func extractVideoID(_ input: String) -> String {
    if input.count == 11 && !input.contains("/") && !input.contains(".") {
        return input
    }
    let patterns = [
        try! NSRegularExpression(pattern: "(?:youtube\\.com/watch\\?v=|youtu\\.be/)([a-zA-Z0-9_-]{11})", options: []),
        try! NSRegularExpression(pattern: "youtube\\.com/embed/([a-zA-Z0-9_-]{11})", options: []),
    ]
    let nsRange = NSRange(input.startIndex..<input.endIndex, in: input)
    for pattern in patterns {
        if let match = pattern.firstMatch(in: input, options: [], range: nsRange),
           let range = Range(match.range(at: 1), in: input) {
            return String(input[range])
        }
    }
    return input
}

private func formatTime(_ seconds: Double) -> String {
    let total = Int(seconds)
    let h = total / 3600
    let m = (total % 3600) / 60
    let s = total % 60
    if h > 0 {
        return String(format: "%d:%02d:%02d", h, m, s)
    }
    return String(format: "%d:%02d", m, s)
}
